bigbluebutton/greenlight
# GLv3-3202: Allow user to "Sign Up".

Backend:
		Done:
                        - Set the API endpoint for the signup '/api/v1/users.json'.
                        - Made some changes to Rubocop.
                        - Added a request spec for the users API and prepared the tests for the signup with/without valid user data.
                        - Improved routing compatibility between SSR (Server side routing) and CSR (Client side routing).
                        - Tried defining an approach for error and request handling (Still require more work).
                        - Improved API resiliency by introducing a much stricter API:
                        -> API calls has to have an explicit `.json` format along with an `Accept: 'application/json'` header.
                        -> The TODOs were left as comments in the code.
	Frontend:
		        - Set a new project structure for the front end JavaScript code influenced by the
		        react-router-demo https://stackblitz.com/edit/github-agqlf5/ :
		            -> main.jsx: An entry point where Routes, Layouts, Global state is passed/defined.
		            -> app.jsx: The parent route component for the `/` endpoint.
		            -> components/: A directory where custom components resides.
		            -> components/styling: A directory where custom components with no business logic resides (i.e. custom buttons).
		            -> components/forms: A directory where custom form components resides.
		            -> routes: A directory where sub routes resides.
		            
		        - Created a Signup route that renders `routes/Signup.jsx` for the `/signup` endpoint.
		        - Created a `components/forms/SignupForm.jsx` form component that describes the signup form.
		        - Created a `helpers/forms/SignupFormHelpers.jsx` module to bundle the helpers for the signup form component.
		        - Created a `components/forms/FormControl.jsx` component that wraps/connects a react bootstrap form control component with the react hook form API.
		        - Created a `comonents/styling/buttons/ButtonLink.jsx` component that wraps/connects a react bootstrap button component with the react router API.
                        
                        TODO: 
                             - Use axios and react-query instead of the fetch API.
                             - Introduce a way to check for email uniqueness asynchronously from the client side.
                             - Improve the form performance.
                             - Improve API error handling. 
                             - Use typescript.
                             - Add any required tests.
                             - Configure the eslinter.




## Description

Allow users to signup in GLv3.

## Testing Steps

1. Pull the code.
2. Install the dependencies `bundle install && npm|yarn install`.
3. Run any left migration `rails db:migrate`
4. Create any left database for testing `rails db:create`
5. Run the linter and tests `bundle exec rubocop --parallel && bundle exec rspec`
6. Run the rails s `rails s` or better `./bin/dev` to run the assets builders processes at once.

To test the API:
    1. Make a POST request to `/api/v1/users.json`.
     1.1 The request has to have an `Accept: application/json` header.
     1.2. The request body should have a JSON representation with this format: 
     `{ user: { :name, :email, :password, :password_confirmation } }` .
     -> When not specifying the `.json` format the API should render back the `components#index` so that CSR handles the request.
     -> When not specifying the `Accept` header correctly the same scenario should occur.
     -> For unknown routes in the server side the CSR should take charge of handling it.
     -> Not respecting the user data required format will return a 400 BAD request with a JSON object in the response body having an errors array with a message **Invalid or Missing parameters.**
     -> Any unpermitted values given in the user data will be filtered, only `name, email ,password,...` will pass.
     
  To test the UI:
    1. Enter the index `/`
    2. You should have this UI: 
![image](https://user-images.githubusercontent.com/29759616/159025132-438a4310-7f8f-4d57-9bce-95687ee1d627.png)

    3. For any other URL except for `"/" and "/signup"`, the UI should render this:    
![image](https://user-images.githubusercontent.com/29759616/159025450-f9e4f063-1347-4997-a4c3-bff871f559d3.png)

4. Entering "/signup" or clicking the `Signup` button should render this:
![image](https://user-images.githubusercontent.com/29759616/159024197-5ee2b245-7e80-4614-9f26-ce62ea150680.png)

5. Clicking the `Home` button should have the same outcome as accessing the index `/` page.
6. Fill the form, validations should run as expected.
7. Data won't get submitted only if the form is valid.
8. The result of the API call can be viewed with the dev tools along with the puma server logs and the users table in the db.


## Screenshots (if appropriate):
